prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+
+#include <fi_util.h>
+
+#include <inttypes.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#define OFI_VERBS_UD_MAJOR_VERSION 1
+#define OFI_VERBS_UD_MINOR_VERSION 5
+
+#define OFI_VERBS_UD_PRIMARY_CAPS (FI_MSG | FI_RMA |	\
+				   FI_SEND | FI_RECV)
+
+#define OFI_VERBS_UD_CAPS OFI_VERBS_UD_PRIMARY_CAPS
+
+#define VERBS_UD_MSG_IOV_LIMIT	256
+#define VERBS_UD_DEF_CQ_SIZE	1024
+
+#define OFI_VERBS_UD_MR_IOV_LIMIT	1
+
+#define VERBS_IB_FABRIC_PREFIX	"IB-0x"
+#define VERBS_IWARP_FABRIC	"Ethernet-iWARP"
+#define VERBS_ANY_FABRIC	"RDMA-Fabric"
+
+#define OFI_VERBS_UD_ANY_SERVICE	0
+
+/*
+ * useful macros
+ */
+#ifndef likely
+#define likely(x) __builtin_expect((x), 1)
+#endif
+#ifndef unlikely
+#define unlikely(x) __builtin_expect((x), 0)
+#endif
+
+/* Helper macro */
+#define OFI_FID_OPS(_fid) (*_fid)->fid.ops
+#define OFI_OPS(_fid) (*_fid)->ops
+
+#define OFI_VERBS_UD_GET_INSTANCE_OF_OBJ(ofi_obj_name,				\
+					 inst)					\
+do {										\
+	inst = container_of(ofi_obj_name##_fid,					\
+			    struct ofi_verbs_ud_##ofi_obj_name,			\
+ 			    util_##ofi_obj_name.ofi_obj_name##_fid);		\
+} while (0)
+
+#define OFI_VERBS_UD_GET_INSTANCE_OF(ofi_obj_name,				\
+				     inst)					\
+do {										\
+	inst = container_of(ofi_obj_name##_fid,					\
+			    struct ofi_verbs_ud_##ofi_obj_name,			\
+ 			    util_##ofi_obj_name.ofi_obj_name##_fid.fid);	\
+} while (0)
+
+#define VERBS_UD_MR_LOCAL(info)							\
+	((FI_VERSION(info->fabric_attr->api_version, FI_VERSION(1,5)) &&	\
+	  (info->mode & FI_LOCAL_MR)) ||					\
+	 (info->domain_attr->mr_mode & FI_MR_LOCAL))
+/* ~Helper macro */
+
+#define VERBS_UD_ATTR_INIT					\
+/* Notes: */							\
+/* - inject_size, size, iov_limit, rma_iov_limit */		\
+/*   are assigned at run-time in @ofi_verbs_ud_info_init */	\
+static struct fi_tx_attr ofi_verbs_ud_tx_attr = {		\
+	.caps = FI_MSG | FI_SEND,				\
+	.comp_order = FI_ORDER_STRICT,				\
+	.iov_limit = VERBS_UD_MSG_IOV_LIMIT,			\
+	.rma_iov_limit = VERBS_UD_MSG_IOV_LIMIT			\
+};								\
+								\
+/* Notes: */							\
+/* - size, iov_limit are assigned at run-time */		\
+/*   in @ofi_verbs_ud_info_init */				\
+static struct fi_rx_attr ofi_verbs_ud_rx_attr = {		\
+	.caps = FI_MSG | FI_RECV,				\
+	.comp_order = FI_ORDER_STRICT,				\
+	.iov_limit = VERBS_UD_MSG_IOV_LIMIT			\
+};								\
+								\
+/* Notes: */							\
+/* - name is assigned at run-time in @ofi_verbs_ud_info_init */	\
+static struct fi_fabric_attr ofi_verbs_ud_fabric_attr = {	\
+	.prov_version = FI_VERSION(OFI_VERBS_UD_MAJOR_VERSION,	\
+				   OFI_VERBS_UD_MINOR_VERSION)	\
+};								\
+								\
+/* Notes: */							\
+/* - cq_cnt, ep_cnt, max_ep_tx|rx_ctx, tx|rx_ctx_cnt */		\
+/*   are assigned at run-time in @ofi_verbs_ud_info_init */	\
+static struct fi_domain_attr ofi_verbs_ud_domain_attr = {	\
+	.threading		= FI_THREAD_SAFE,		\
+	.control_progress	= FI_PROGRESS_AUTO,		\
+	.data_progress		= FI_PROGRESS_AUTO,		\
+	.resource_mgmt		= FI_RM_DISABLED,		\
```

Comment:
Ok. I completed the implementation. Now this option should work. I updated capabilities too.